File: solutions/elixir/collatz-conjecture/1/lib/collatz_conjecture.ex
defmodule CollatzConjecture do
  @doc """
  calc/1 takes an integer and returns the number of steps required to get the
  number to 1 when following the rules:
    - if number is odd, multiply with 3 and add 1
    - if number is even, divide by 2
  """
  @spec calc(input :: pos_integer()) :: non_neg_integer()
  def calc(input, steps \\ 0)

  def calc(input, _steps) when not is_integer(input) or input <= 0 do
    raise FunctionClauseError
  end

  def calc(1, steps) do
    steps
  end

  def calc(current, steps) when rem(current, 2) == 0 do
    calc(div(current, 2), steps + 1)
  end

  def calc(current, steps) do
    calc(3 * current + 1, steps + 1)
  end
end

File: solutions/erlang/collatz-conjecture/1/src/collatz_conjecture.erl
-module(collatz_conjecture).

-export([steps/1]).

% public function should only allow positive numbers greater than 0
steps(N) when (N>0) and is_integer(N) -> steps(N,0);
steps(_N) -> erlang:error(badarg).

% Private function which accumulates the steps and does the division
% by 2 if even and multiplies by 3 and adds one if odd
steps(N, Acc) when N==1 -> Acc;
steps(N, Acc) when (N rem 2) ==0 -> steps(N div 2, Acc+1);
steps(N, Acc) -> steps(((N*3)+1), Acc+1).
